sysfs/rapl.go

```diff
@@ -0,0 +1,139 @@
+type RaplValue struct {
+	Name           string // name of RAPL value from file "name"
+	Index          int    // index (different value for duplicate names)
+	Microjoules    uint64 // RAPL microjoule value
+	MaxMicrojoules uint64 // max RAPL microjoule value
+}
+
+// internal raplValue which contains also the sysfs filename
+type raplValue struct {
+	sysfsFilename string
+	RaplValue
+}
+
+// RaplReader reads RaplValues from sysfs
+type RaplReader struct {
+	raplValues []raplValue
+}
+
+// NewRaplReader returns a new reader for RAPL values.
+// When RAPL files are not present, returns nil with error
+// https://www.kernel.org/doc/Documentation/power/powercap/powercap.txt
+func NewRaplReader(fs FS) (*RaplReader, error) {
+	reader := RaplReader{}
+	raplDir := fs.sys.Path("class/powercap")
+
+	files, err := ioutil.ReadDir(raplDir)
+	if err != nil {
+		return nil, errors.New(
+			"no sysfs powercap / RAPL power metrics files found")
+	}
+
+	// count name usages to avoid duplicates (label them with an index)
+	countNameUsages := make(map[string]int)
+
+	// loop through directory files searching for file "name" from subdirs
+	for _, f := range files {
+		nameFile := filepath.Join(raplDir, f.Name(), "/name")
+		nameBytes, err := ioutil.ReadFile(nameFile)
+		if err == nil {
+			// add new rapl value since name file was found
+			name := strings.TrimSpace(string(nameBytes))
+
+			// make RAPL name prometheus-compatible
+			modifiedPromName := strings.Replace(name, "-", "_", -1)
+
+			// get a pair of index and final name
+			index, modifiedPromName := getIndexAndName(countNameUsages,
+				modifiedPromName)
+
+			maxMicrojouleFilename := filepath.Join(raplDir, f.Name(),
+				"/max_energy_range_uj")
+			microjouleFilename := filepath.Join(raplDir, f.Name(),
+				"/energy_uj")
+			maxMicrojoules, err := util.ReadUintFromFile(maxMicrojouleFilename)
+			if err != nil {
+				return nil, err
+			}
+
+			rv := RaplValue{
+				Name:           modifiedPromName,
+				Index:          index,
+				Microjoules:    0,
+				MaxMicrojoules: maxMicrojoules,
+			}
+			privRv := raplValue{
+				sysfsFilename: microjouleFilename,
+				RaplValue:     rv,
+			}
+
+			reader.raplValues = append(reader.raplValues, privRv)
+
+			// Store into map how many times this name has been used. There can
+			// be e.g. multiple "dram" instances without any index postfix. The
+			// count is then used for indexing
+			countNameUsages[modifiedPromName] = index + 1
+		}
+	}
+
+	return &reader, nil
+}
+
+// NewRaplValues returns a slice of RaplValues read from the sysfs
+// https://www.kernel.org/doc/Documentation/power/powercap/powercap.txt
+func (reader RaplReader) NewRaplValues() ([]RaplValue, error) {
+	length := len(reader.raplValues)
+	retValues := make([]RaplValue, length, length)
+	var err error
+	for i := 0; i < length; i++ {
+		reader.raplValues[i].Microjoules, err =
+			util.ReadUintFromFile(reader.raplValues[i].sysfsFilename)
```

Comment:
This error is not being checked each time through the loop, so if you get an error on first iteration and then a `nil` on the second, the error will be ignored, and only the `nil` will be returned.

Reply:
true, I'll fix that